File: abc/abc269/src/bin/e.py
def main():
    N = int(input())
    head = 0
    tail = N
    mid = (head + tail) // 2
    while tail - head > 1:
        print(f"? {head + 1} {mid} 1 {N}", flush=True)

        if mid - head > int(input()):
            head = head
            tail = mid
        else:
            head = mid
            tail = tail
        mid = (head + tail) // 2
    ansx = tail

    head = 0
    tail = N
    mid = (head + tail) // 2
    while tail - head > 1:
        print(f"? 1 {N} {head + 1} {mid}", flush=True)

        if mid - head > int(input()):
            head = head
            tail = mid
        else:
            head = mid
            tail = tail
        mid = (head + tail) // 2
    ansy = tail

    print(f"! {ansx} {ansy}", flush=True)


if __name__ == "__main__":
    main()

File: abc/abc269/src/bin/e.rs
use std::io::{stdin, stdout, BufReader, Write};

use proconio::{input, source::line::LineSource};

#[allow(non_snake_case)]
fn main() {
    let stdin = stdin();
    let mut source = LineSource::new(BufReader::new(stdin.lock()));
    input! {
        from &mut source,
        N: usize,
    }

    let mut head = 0;
    let mut tail = N;
    let mut mid = (head + tail) / 2;

    while tail - head > 1 {
        println!("? {} {} {} {}", head + 1, mid, 1, N);

        input! {
            from &mut source,
            x: usize,
        }

        if mid - head > x {
            tail = mid;
        } else {
            head = mid;
        }
        mid = (head + tail) / 2;
    }
    let ansx = tail;

    let mut head = 0;
    let mut tail = N;
    let mut mid = (head + tail) / 2;

    while tail - head > 1 {
        println!("? {} {} {} {}", 1, N, head + 1, mid);

        input! {
            from &mut source,
            x: usize,
        }

        if mid - head > x {
            tail = mid;
        } else {
            head = mid;
        }
        mid = (head + tail) / 2;
    }
    let ansy = tail;

    println!("! {} {}", ansx, ansy);
    stdout().flush().unwrap();
}
